src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Analysis.cs

```diff
@@ -0,0 +1,90 @@
+            }
+
+            var (syntaxKinds, analyzeSyntax) = GetSyntaxNodeAction();
+            if (analyzeSyntax != null)
+            {
+                context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds);
+            }
+
+            // check for all the other things that can be registered as well.
+        }
+
+        private void AnalyzeIfEnabled(CodeBlockAnalysisContext context, Action<CodeBlockAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext, TOptionKind> analyze,
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
+
+            analyze(context, optionValue.Value);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+        {
+            public readonly TCodeStyleProvider _codeStyleProvider;
+
+            protected DiagnosticAnalyzer(bool configurable = true) 
+                : this(new TCodeStyleProvider(), configurable)
+            {
+            }
+
+            private DiagnosticAnalyzer(TCodeStyleProvider codeStyleProvider, bool configurable)
+                : base(codeStyleProvider._descriptorId, codeStyleProvider._title, codeStyleProvider._message, configurable)
+            {
+                _codeStyleProvider = codeStyleProvider;
+            }
+
+            protected sealed override void InitializeWorker(AnalysisContext context)
+                => _codeStyleProvider.DiagnosticAnalyzerInitialize(context);
+
+            public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+                => _codeStyleProvider.GetDiagnosticAnalyzerCategory();
+
+            public sealed override bool OpenFileOnly(Workspace workspace)
+                => _codeStyleProvider.DiagnosticsForOpenFileOnly(workspace);
```

Comment:
all members in these classes are intentionally sealed. All logic should be provided in the CodeStyleProvider subclass.